messenger/src/redis_messenger.rs

```diff
@@ -95,56 +102,51 @@ impl Messenger for RedisMessenger {
         };
 
         // Put serialized data into Redis.
-        let result: RedisResult<()> = self.connection.as_mut().unwrap().xadd_maxlen(
-            stream_key,
-            maxlen,
-            "*",
-            &[(DATA_KEY, bytes)],
-        );
-
-        // Log but do not return errors.
+        let result: RedisResult<()> = self
+            .connection
+            .as_mut()
+            .unwrap()
+            .xadd_maxlen(stream_key, maxlen, "*", &[(DATA_KEY, &bytes)])
+            .await;
+
         if let Err(e) = result {
             error!("Redis send error: {e}");
+            return Err(GeyserPluginError::Custom(Box::new(
+                MessengerError::SendError { msg: e.to_string() },
+            )));
         } else {
             info!("Data Sent");
         }
 
         Ok(())
     }
 
-    fn recv(&mut self) -> Result<()> {
+    async fn recv(&mut self, stream_key: &'static str) -> Result<Vec<(i64, &[u8])>> {
         let opts = StreamReadOptions::default()
-            .block(1000)
-            .count(100000)
+            .block(0) // Block forever.
+            .count(1) // Get one item.
             .group(GROUP_NAME, CONSUMER_NAME);
 
-        // Setup keys and ids to read on all configured streams.
-        let keys: Vec<&str> = self.streams.keys().map(|s| *s).collect();
-        let ids: Vec<&str> = vec![">"; keys.len()];
-
-        // Read on all streams and save the reply. Log but do not return errors.
-        match self
+        // Read on stream key and save the reply. Log but do not return errors.
+        self.stream_read_reply = match self
             .connection
             .as_mut()
             .unwrap()
-            .xread_options(&keys, &ids, &opts)
+            .xread_options(&[stream_key], &[">"], &opts)
+            .await
         {
-            Ok(reply) => self.stream_read_reply = reply,
-            Err(e) => error!("Redis receive error: {e}"),
-        }
-
-        Ok(())
-    }
+            Ok(reply) => reply,
+            Err(e) => {
+                error!("Redis receive error: {e}");
+                return Err(GeyserPluginError::Custom(Box::new(
+                    MessengerError::ReceiveError { msg: e.to_string() },
+                )));
+            }
+        };
```

Comment:
What happens to messages that fail to process? This can vary dramatically by system  and configuration, and I'll have to read up on Redis Streams.

Reply:
Right now we will ignore these errors and just the tests will fail.  Created issue to track this design work:
https://app.zenhub.com/workspaces/protocol-engineering-61eebb497fb4920012651e19/issues/metaplex-foundation/metaplex-program-library/537